### PR TITLE
NLP Structured Phrase

### PR DESCRIPTION
**Pull request by Dtphelan1**
Initial implementation of the nlp mechanism of structured phrases. This PR introduces the #nlptoxicity shortcut and the new `NLPHashtag` shortcut type. The presence of an `NLPHashtag` activates the NLPHashtagPlugin, which looks for a stop character ('.' or '?' or '!') followed by an end character (space or newline). If these characters are found, the plugin sends of the sentence between the `NLPHashtag` these endOfPhraseCharacters to the NLPEngine that Sam Bayer has built (currently this sever is hosted on my machine; I'm in the process of setting it up on the big mac in the lab). This sentence gets parsed and FluxNotes gets the result of this parsing. Should the engine identify phrases that we need to insert, we will find the text in our editor corresponding to each phrase and replace it with structuredPhrase for the appropriate child shortcut. The `ToxicityGradeCreator, ToxicityAdverseEventCreator, and ToxicityAttributionCreator` shortcuts all have had metadata added to them in shortcuts.json in order to specify how to parse the data returned back from the NLPEngine to retrieve the associated shortcut triggers.

## Notes 
- Currently, this insertion can insert in the wrong location if the sentence passed off to the NLPEngine contains a structured phrase (e.g. parsing `anything @age[[51]] grade 1` will insert a structured phrase for grade 1, but in the wrong location and simultaneously fail to delete `grade 1`). This is a known bug and is being tracked.

- If the NLP processing takes a noticable amount of time, the phrase "NLP Processing..." appears in the upper-right hand of the breadcrumbs menu. 

- While making changes to FluxNotesEditor I also ended up breaking off the rendering of note description content into its own function, aptly names `renderNoteDescriptionContent`.

- I also introduce a new function for inserting structured phrases that allow users to specify a range and insert there directly. The plain insertStructuredPhrase function still works, but calls this more specific function, passing the current editor selection as the range argument.

- **N.B.** This PR does not remove #toxicity, it simply adds a second version of the toxicity shortcut that changes the same objects on the patient JSON, but stands independently from #toxicity in all other facets (e.g. autocomplete, contextOptions suggestions, etc...)

- Note Parser has not been updated to interact with the NLPEngine. There is a JIRA to address this, and @gregquinn2001 has agreed with me to put this off until then.

_Pull requests into Flux require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:._

## Submitter:

**Running the Application**

- [x] Manually tested in Chrome
- [x] Manually tested in IE

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- **N/A** Cheat sheet is updated
- **N/A** Demo script is updated 
- **N/A** Documentation on Wiki has been updated 
- **N/A** Additional technical components and libraries are documented and added to wiki as needed

**NoteParser**

- **N/A** Note parser has been updated if structured phrases change

**Code Quality**

- [x] 4-space indents - convert any tabs to spaces
- [x] Use public class field syntax for binding functions - ex. handleChange = () => { ... };
- [x] Code is commented

**Tests**

- [x] Existing tests passed
- **N/A** Added Enzyme-UI tests for slim mode 
- **N/A** Added Enzyme-UI tests for full mode
- **N/A** Added backend tests for new functionality added
- **N/A** Added note parser examples to test new functionality

## Reviewer 1: Name

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

## Reviewer 2: Name

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: src/notes/NLPHashtagPlugin.jsx [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+function createOpts(opts) {
+	opts = opts || {};
+	return opts;
+}
+
+function SingleHashtagKeywordStructuredFieldPlugin(opts) {
+    opts = createOpts(opts);
+    const shortcutManager = opts.shortcutManager;
+    const contextManager = opts.contextManager;
+	const createShortcut = opts.createShortcut;
```
Comment (jafeltra) [inline src/notes/NLPHashtagPlugin.jsx]: This is a small code styling thing, but it seems like there is a mix of tabs and spaces in this file.

Comment (Dtphelan1) [reply src/notes/NLPHashtagPlugin.jsx]: Good catch

**Comment by jafeltra** [comment]
For your first bullet of notes, should we add a Jira to track the incorrect insertion when other structured phrases are used?

**Comment by Dtphelan1** [comment]
Yes! I've just added it 👍

**Comment by jafeltra** [comment]
This looks awesome! I talked to Dylan about some issues we had with the NLP engine running on the mac, so I can relay the fix we made to anyone else reviewing this PR. Now that I have it running, I did hit one error that I think is unexpected.

[JA1] When hitting enter to end the sentence and trigger the NLP engine, I got an error.
The text I typed: #nlptoxicity grade five nausea disease.
The error I got:
NLPHashtagPlugin.jsx:241 Error: [setAttributeValue]Unsupported attribute attribution for context shortcut InsertValue
    at InsertValue.setAttributeValue (Context.jsx:64)
    at CreatorChild.setText (CreatorChild.jsx:118)
    at CreatorChild.initialize (CreatorChild.jsx:51)
    at FullApp._this.newCurrentShortcut (FullApp.jsx:202)
    at parseArryOfPhrases (NLPHashtagPlugin.jsx:159)
    at parsePhrases (NLPHashtagPlugin.jsx:192)
    at processNLPEngineData (NLPHashtagPlugin.jsx:217

**Comment by danlee1025** [comment]
[DL1]  This may be related to the bug noted in the PR.  When I type `#nlptoxicity grade 1 nausea. `  It will insert `#nlptoxicity #grade 1 #nausea. `  Now if I immediately type `treatment. ` it inserts #treatment but does not delete treatment.  The whole text looks like `#nlptoxicity #treatment #grade 1 #nausea. treatment. `

**Comment by Dtphelan1** [comment]
DL1 is already noted.

**Comment by Dtphelan1** [comment]
JA1 is happening due to the fact that, after hitting a newline, we enter a new block of text which means the previous context #nlptoxicity is no longer valid. That explains why we get that bug. I'm not sure what we want to do to solve it... Anyone have any ideas on how we could avoid that problem?

**Comment by jafeltra** [comment]
This looks good to me! We made a separate jira for JA1 so I think that should be set. The only other error I came across was this:

[JA2]: When using #nlptoxicity, if you end a sentence so that the NLP is processing, and while it is, delete a piece of the data that would get structured, I get an error: 
```
Uncaught (in promise) TypeError: Cannot read property 'marks' of undefined
    at Document.getMarksAtRangeAsArray (node.js:900)
    at Document.object.(anonymous function) [as getMarksAtRangeAsArray] (http://localhost:3000/static/js/bundle.js:202725:29)
    at Document.getMarksAtRange (node.js:861)
    at Document.object.(anonymous function) [as getMarksAtRange] (http://localhost:3000/static/js/bundle.js:202725:29) ...
```
I'm not sure if this is something we'd want to support, but maybe just have it error more gracefully. If we want to make it a separate Jira as well, I'm okay with that!

Other than that, I tested inserting other shortcuts and text into the app, and everything seems to be working as expected. This is really cool!

**Comment by jafeltra** [comment]
Made a Jira for JA2!

**Comment by gregquinn2001** [comment]
[GQ1] I get two grade 1's. 

@condition[[ductal carcinoma of breast]] #nlptoxicity blah blah grade 1 level. 

I end up with two #grade 1 structured phrases at end replacing 'grade 1 level'.

**Comment by jafeltra** [comment]
Checked this again after the rebase and it looks good! I'm going to merge it in!

**Pull request merged**